test-gui/src/package/view-decoded-linear-position-plot/DecodedLinearPositionDrawing.ts

```diff
@@ -0,0 +1,212 @@
+import { useCallback, useMemo, useRef } from 'react'
+import { inferno, magma, plasma, viridis } from 'scale-color-perceptual'
+import { ValidColorMap } from '../view-track-position-animation/TPADecodedPositionLayer'
+import { TimeColumn } from './DecodedLinearPositionLineRepresentations'
+
+// const DARK_DOVE_GREY = 'rgb(119, 118, 114)'
+
+export const usePositions = (maxOffscreenCanvasHeight: number, positionsKey: number[]) => {
+    return useMemo(() => {
+        const basePosition = positionsKey[0] // bin coordinates are actually the top of the range
+        const lastPosition = (positionsKey.at(-1) ?? basePosition) + basePosition
+        const vscale = Math.floor(maxOffscreenCanvasHeight/lastPosition)
+        const canvasPositions = [0, ...positionsKey, lastPosition].map(n => vscale * n)
+        const targetHeight = Math.ceil(vscale * lastPosition)
+        return { canvasPositions, targetHeight }
+    }, [positionsKey, maxOffscreenCanvasHeight])
+}
+
+
+export type OffscreenRenderProps = {
+    canvas: HTMLCanvasElement | undefined
+    canvasTargetWidth: number
+    canvasTargetHeight: number
+    painter: OffscreenPainter
+    linesRepresentation: TimeColumn[]
+    scale: number
+    downsampledRangeStart: number
+    downsampledRangeEnd: number
+}
+// Given a Canvas with stuff drawn on it and a data set, and a desired range of data from the data set,
+// return the x-coordinate pixel range in the offscreen canvas that contains the requested data range.
+export const useOffscreenCanvasRange = (props: OffscreenRenderProps): [number, number] => {
```

Comment:
This is mostly the logic for indexing into the offscreen canvas. (Actually updating it has, happily, been moved to the function below, which makes it a bit easier to follow.)

Apart from some manipulation of the canvas size here--which we don't really need to do--this could probably be separated entirely from the actual data, but I haven't done that yet.

Also, a lot of this seems like it ought to be handled with hooks, but the evaluation order wasn't high enough--the rest of the drawing operations happen even if values they depend upon should have been changed by a hook. There's probably a better way to write it, but I haven't figured it all out yet.

Anyway, the overall plan is that you want to make sure the canvas is the right size, invalidate the cache if it isn't or if the scale has changed, and then (if the visible data is all already on the offscreen canvas) return the indices that should be copied.

If not, we need to update the offscreen cache. Do that so that the midpoint of the visible area will be the midpoint of the (hopefully larger) cache, then update our record of the timepoint range that's present in the offscreen cache, and then finally return the indices of the desired viewpoint based on the new cache contents.

Note: it's a rule that 1 x pixel in the offscreen canvas cache corresponds to 1 time point at the current scale factor.